MAINT: Simplify class hierarchy of the parsing unit formats

mhvk opened the pull request:
This rather large refactor was triggered by @eerovaher's [comment in #17374](
https://github.com/astropy/astropy/pull/17374#issuecomment-2473973198) that it should become possible to get of the private `_to_string()` method. This was indeed possible, with a relatively minor refactor. 

However, it seemed sensible to go a bit further, and make the whole class hierarchy a bit more logical. The rationale is below, split by commit. If helpful, I can limit this PR to commits 1 and 2. Opening as draft in part because my naming is not great, in part because some things are a bit hacky (like not registering classes ending in `Base`), and in part because I'm not sure about the inheritance structure (use mixins?). Also, one definitely needs to do something more sensible with `format_exponential_notation`... Finally, while I think all classes we use should be documented, it is not clear we really want to commit to the in-between classes - there really are a few too many little private methods to fine-tune results...

The current state is the following inheritance diagram (classes with * implement a parser):
```
   Base⭢Console⭢Unicode
       ︳       ⮡Latex⭢LatexInline
       ⮡Generic*⭢FITS⭢OGIP*
                      ⮡CDS*
                      ⮡VOUnit
```
Logically, OGIP and CDS should not inherit from Generic, since they have different parsers, while VOUnit probably should inherit directly from Generic, since it shares little with FITS.

Next are ~four~ five commits (EDIT: updated with move to mixin classes): 

1. Remove the dependence on FITS. This needs duplicating `_parse_unit` since the one from Generic is no good (but those methods get removed later). One now has:
```
   Base⭢Console⭢Unicode
       ︳       ⮡Latex⭢LatexInline
       ⮡Generic*⭢FITS
                 ⮡OGIP*
                 ⮡CDS*
                 ⮡VOUnit
```
2. Removes `_to_string()`, merging the code in `to_string()`.
3. Make all parsing units depend on a single base.
```
   Base⭢Console⭢Unicode
       ︳       ⮡Latex⭢LatexInline
       ⮡ParsingFormatBase⭢Generic*⭢FITS
                          ︳       ⮡VOUnit
                          ⮡OGIP*
                          ⮡CDS*
```
4. Factor out the generic parser, so Generic, FITS, and VOUnit can just add their own stuff around it. `_parse_unit` in the latter two can now be removed again.
```
   Base⭢Console⭢Unicode
       ︳       ⮡Latex⭢LatexInline
       ⮡ParsingFormatBase⭢GenericParserBase*⭢Generic
                          ︳                 ⮡FITS
                          ︳                 ⮡VOUnit
                          ⮡OGIP*
                          ⮡CDS*
```
5. (EDIT: added later) Move to mixin classes so that we can keep those private implementation details for now.

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

nstarman commented:
This is currently in draft mode. Do you want us to wait to review?

mhvk commented:
@nstarman - yes, please do review! It is in draft mode because I'm not sure about all of my choices (but am quite sure it is better than what we have!); see top post for detail.

nstarman commented on astropy/units/format/base.py:
```diff
@@ -39,6 +43,8 @@ def __init_subclass__(cls, **kwargs):
         # Keep a registry of all formats.  Key by the class name unless a name
         # is explicitly set (i.e., one *not* inherited from a superclass).
         if "name" not in cls.__dict__:
+            if cls.__name__.endswith("Base"):
```
I agree this is janky. I think the way to solve this is instead to check for a class being abstract or not — `inspect.isabstract`. This would of course mean that we make all the base classes (BCs) into ABCs. But I think we should do that!

eerovaher replied:
What abstract classes are supposed to mean is that they have abstract methods the presence of which prevents creating instances directly, but it is possible to instantiate subclasses of abstract classes if they have implementations for all the abstract methods. The `astropy` unit formatters are different because they are never meant to be instantiated, no matter what they implement. Furthermore, the implementation of `Base.__new__()` would conflict with `ABC.__new__()`.

Luckily for us, there already exists a naming convention for things that should not be public:
```suggestion
            if cls.__name__.startswith("_"):
```

nstarman commented on astropy/units/format/base.py:
```diff
@@ -208,3 +207,111 @@ def parse(cls, s: str) -> UnitBase:
+            if str(e):
+                raise
+            else:
+                raise ValueError(f"Syntax error parsing unit '{s}'")
```
Raise from or from None?

```suggestion
                raise ValueError(f"Syntax error parsing unit '{s}'") from e
```

nstarman replied:
Resolved.

nstarman commented on astropy/units/format/base.py:
```diff
@@ -208,3 +207,111 @@ def parse(cls, s: str) -> UnitBase:
+                raise ValueError(f"Syntax error parsing unit '{s}'")
+
+    @classmethod
+    def _parse_unit(cls, unit: str, detailed_exception: bool = True) -> UnitBase:
```
Nitpick.

```suggestion
    def _parse_unit(cls, unit: str, *, detailed_exception: bool = True) -> UnitBase:
```

nstarman commented on astropy/units/format/base.py:
```diff
@@ -208,3 +207,111 @@ def parse(cls, s: str) -> UnitBase:
+        return None
+
+    @classmethod
+    def _validate_unit(cls, unit: str, detailed_exception: bool = True) -> None:
```
Nitpick

```suggestion
    def _validate_unit(cls, unit: str, *, detailed_exception: bool = True) -> None:
```

nstarman reviewed (commented):
I think this is a great step in the right direction. I actually think we should take one step further, which will help solve issues like registration and be very clear and understandable to users. That further step is to use the Abstract/Final pattern where no concrete class can inherit from another concrete class, only ABCs. Consider the following inheritance diagram, which is quite similar to the step 4 result

```
AbstractBase ⭢ AbstractConsole ⭢ Console
                               ⭢ Unicode
                               ⭢ AbstractLatex ⭢ Latex
                                               ⭢ LatexInline
             ⭢ AbstractParsingFormat ⭢ OGIP
                                     ⭢ CDS
                                     ⭢ AbstractGeneric ⭢ Generic
                                                       ⭢ FITS
                                                       ⭢ VOUnit
```

Now it is 100% clear to users which are usable format classes (any one not "Abstract") and how to define their own format (inherit from the relevant ABC). From our perspective it's also 100% clear which classes should be registered (any that aren't `inspect.isabstract`) and this Abstract/Final pattern also protects against many of the tricky bits of `super()` (see ["co-operative multiple inheritance" in here](https://docs.kidger.site/equinox/pattern/)). From a typing perspective we can `@final` all concrete classes, which further helps us and users.

mhvk commented:
@nstarman - I like the idea of adding a separate "base" class also on the console side and am happy to do it either here or in follow-up. A few notes/questions:
1. I don't think we can change the name of `Base` - it has been public and people may have inherited from it. That said, throughout astropy we fairly often use `*Base` effectively as an abstract base class that is not usable itself (e.g., `UnitBase`, `RepresentationBase`, `TimeBase`, though unfortunately also `BaseColumn` and `BaseCoordinateFrame`), so perhaps we can implement your suggestion but follow this naming scheme.
2. Our format classes never get initialized so in that sense they are not a perfect fit for ABCs -- though I see that it is possible to label an `@abstractmethod` as `@classmethod` and a quick experiment shows things work as expected. Is there any worry about people inheriting from `Base`? I guess not as long as we don't make the "wrong" methods abstract.
3. In principle people may have inherited from any of the existing classes, so not sure we can do the `@final`...

I guess my other worry is about setting in stone a set of mostly private methods that I feel could use further simplification (@eerovaher has already gotten us quite far, but really, the chain to, e.g., get a simple unit is a bit long: `_get_unit` -> `_parse_unit` -> `_validate_unit` -> `_units[name]`. But perhaps the best is not to worry too much, and merge something like what I have and then proceed to polish it before it is exposed in 7.1 (ensuring their is no actual API change...). I also do not know if there are actual different format classes out in the wild...

p.s. Separately, I agree with all your smaller comments, but these are for pieces of code that I just happened to move around... I can do them in a separate commit once we agreed on the best way forward generally.

nstarman commented:
> I don't think we can change the name of Base.

Sure. Certainly not for 7.1. 

> so not sure we can do the @final

`@final` is type-time only. IDEs can warn you if you're subclassing a `final` class, but it has no runtime consequences.

mhvk commented:
Good points. The main argument for `*Base` then would remain that we use it throughout astropy. Anyway, somewhat orthogonal to what's needed for 7.1. 

Am tempted to go ahead and implement your abstract scheme with existing names, but will wait for comments from @eerovaher first...

eerovaher commented:
My understanding is that normal users are never supposed to use the unit formatters directly. They are always supposed to use the formatter names instead. The reason the formatters are not private is so that downstream developers could implement new formatters by inheriting from one of ours. This is exactly the opposite of the abstract/final pattern we are discussing. That pattern has many good properties (e.g. if code is compiled with Mypyc then the compiler is allowed to optimize more aggressively by looking up `Final` attributes at compile time), but we are not doing greenfield development here.

mhvk commented:
Agreed that we should keep in  mind that the whole formatter class system is not aimed at regular astropy users, but that the "users" we should consider are people who might want to subclass formatters. 

The question then  is whether we're OK with them subclassing some of the intermediate classes like my `ParsingFormatBase`, or whether they should only subclass `Base` and the existing format classes (which are documented, so free game in that respect).

In principle, I would say, yes, it should be possible and even preferred to subclass those, as they would provide better bases to build on than the final implementations. But I'm not sure we are at that stage yet -- as I mentioned above, I somewhat dislike the large number of private methods that we currently have (all logical, but many implemented just to allow for minor variations that are difficult to deal with in class hierarchy) and would really like to avoid getting stuck to those before we have finished refactoring.

That said, the new intermediate classes I introduced consist essentially completely of private methods -- the only exception is `parse` in `ParsingFormatBase`, which could be moved up to `Base` without trouble (it would involve adding a check on the existence of `_parser` in the `Base` implementation, raising the current error if it does not exist, and calling it if it does). If so, it is quite reasonable to make them private as well... The only issue I see is that when we document classes with `sphinx`, it is a bit strange if the class bases are not documented as well.

Actually, one advantage of moving `parse` up, is that `ParsingFormatBase` does become orthogonal to `Base`, so it could become a mixin. Similarly, `GenericParserBase` does not override anything, so it could become a mixin class too, i.e., one would end up with the following definitions of our concrete format parsing classes:
```
class [Generic, FITS, VOUnit](Base, ParsingFormat, GenericParser):
    ...

class [CDS, OGIP](Base, ParsingFormat):
    ...
```
This has the side benefit of removing the worries about registration. What do you think?

eerovaher commented:
> The question then is whether we're OK with them subclassing some of the intermediate classes like my `ParsingFormatBase`, or whether they should only subclass `Base` and the existing format classes (which are documented, so free game in that respect).
> 
> In principle, I would say, yes, it should be possible and even preferred to subclass those, as they would provide better bases to build on than the final implementations. But I'm not sure we are at that stage yet -- as I mentioned above, I somewhat dislike the large number of private methods that we currently have (all logical, but many implemented just to allow for minor variations that are difficult to deal with in class hierarchy) and would really like to avoid getting stuck to those before we have finished refactoring.

All the classes we add here should be private, at least for now, so that we could continue with refactoring without having to worry about the public API.

> That said, the new intermediate classes I introduced consist essentially completely of private methods -- the only exception is `parse` in `ParsingFormatBase`, which could be moved up to `Base` without trouble (it would involve adding a check on the existence of `_parser` in the `Base` implementation, raising the current error if it does not exist, and calling it if it does).

We have to be careful with `Base.parse()`, but that idea sounds like it could work.

> Actually, one advantage of moving `parse` up, is that `ParsingFormatBase` does become orthogonal to `Base`, so it could become a mixin. Similarly, `GenericParserBase` does not override anything, so it could become a mixin class too, i.e., one would end up with the following definitions of our concrete format parsing classes:
> 
> ```
> class [Generic, FITS, VOUnit](Base, ParsingFormat, GenericParser):
>     ...
> 
> class [CDS, OGIP](Base, ParsingFormat):
>     ...
> ```

`GenericParserMixin` could just inherit from `ParsingFormatterMixin` so that `FITS` and others wouldn't have to inherit from two different mixins.

> This has the side benefit of removing the worries about registration. What do you think?

Private mixin classes that only provide private methods sounds like it might be a good solution, as long as we are careful with how we handle the public `parse()` method.

mhvk commented:
OK, this sounded like a plan, so I implemented it -- turned out to be remarkably painless (see last commit; this despite the multiple pushes as I had forgotten to undo some things).

I like that this way we do not preclude a final goal along the lines of what @nstarman suggested.

nstarman commented on astropy/units/format/base.py:
```diff
@@ -203,8 +200,135 @@ def _to_string(cls, unit: UnitBase, *, fraction: bool | str) -> str:
 
     @classmethod
-    def parse(cls, s: str) -> UnitBase:
+    def parse(cls, s: str, debug: bool = False) -> UnitBase:
```
Nitpick
```suggestion
    def parse(cls, s: str, *, debug: bool = False) -> UnitBase:
```

eerovaher replied:
I don't think we should be changing the signature of a public method at all in this pull request.

mhvk replied:
Yes, agreed, this is better as just a refactoring.

nstarman replied:
SGTM. For a followup.

nstarman replied:
e.g. https://docs.astral.sh/ruff/rules/boolean-type-hint-positional-argument/

eerovaher replied:
The `Base.parse()` method mostly exists to define the public API for downstream packages. Adding a new parameter changes that API and we definitely shouldn't do that in what is supposed to be a refactoring of the private API.

nstarman replied:
IDK. All the other `parse` methods have this optional kwarg. Code shouldn't break for downstream users.

nstarman commented on astropy/units/format/base.py:
```diff
@@ -203,8 +200,135 @@ def _to_string(cls, unit: UnitBase, *, fraction: bool | str) -> str:
+    _deprecated_units: ClassVar[frozenset[str]] = frozenset()
+
+    @classmethod
+    def _parse_unit(cls, unit: str, detailed_exception: bool = True) -> UnitBase:
```
Nitpick

```suggestion
    def _parse_unit(cls, unit: str, *, detailed_exception: bool = True) -> UnitBase:
```

nstarman replied:
pushing this to a future PR?

nstarman commented on astropy/units/format/base.py:
```diff
@@ -203,8 +200,135 @@ def _to_string(cls, unit: UnitBase, *, fraction: bool | str) -> str:
+        return None
+
+    @classmethod
+    def _validate_unit(cls, unit: str, detailed_exception: bool = True) -> None:
```
```suggestion
    def _validate_unit(cls, unit: str, *, detailed_exception: bool = True) -> None:
```

nstarman reviewed (commented):
Looks good! I agree this is the right approach. I just want to confirm that the mixing are completely orthogonal.

eerovaher reviewed (commented):
Introducing the private mixin classes that only implement private methods looks like a good idea, but making any changes to the public `Base.parse()` method should be avoided. See e.g. https://github.com/astropy/astropy/blob/7f90edbb7d5b7f5d3116040e825affd7d99bc775/astropy/units/format/__init__.py#L59-L73
That function is used for constructing an error message which lists parsing formatters separately from the write-only formatters. Importantly for this pull request, the division into the two groups is done based on their `parse()` method. Although it is quite dynamical, it does make sense on current `main` because no parsing formatter can use `Base.parse()` and no write-only formatter has any reason to implement its own. The changes being proposed here would allow a parsing formatter to work with a `parse()` implementation inherited from `Base`, which can cause misleading error messages. Perhaps it is best to keep the `_do_parse()` method around, at least for now.

mhvk commented:
See https://github.com/astropy/astropy/pull/17397/commits/02fd18e3c34f168ac45be8f45ea3a2d141827cf1

eerovaher commented:
I stumbled upon `_known_formats()` in time, but I can't be fully confident that there aren't any other functions that might be using `Base.parse()` in ways we haven't accounted for. I would prefer to not touch `Base.parse()` at all in this pull request.

eerovaher commented on astropy/units/tests/test_format.py:
```diff
@@ -1110,3 +1110,15 @@ def test_Fits_name_deprecation():
     assert Fits is u.format.FITS
+
+
+def test_to_string_unknown_format():
+    # Check full message to ensure we correctly classify in-out and
```
You've already written a test for this: https://github.com/astropy/astropy/blob/28d1279744521d5d98e56b76f7f1f0259e2bcc41/astropy/units/tests/test_format.py#L1034-L1036

mhvk replied:
OK, clearly I should not just have looked for tests with `.to_string()`...  I now combined the two tests in a new class, with both checking the full message.

mhvk commented:
On `.parse()`, my sense is to keep the new logic. Really, the code in `__init__` was not great - relying on what logically should be an implementation detail. Indeed, even my change was not great: it should just try to parse something. So, I changed that now...

mhvk commented:
Well, so much for just doing a try/except on `format_.parse("m")` - that turns out to be re-entrant and just mess things up completely. This is more fragile than it should be...  Anyway, I went back to what I had for now...

I still think it makes little sense to duplicate `.parse` -- if we go that route, I think we should just give up on the mixin idea, and go back to subclassing instead.

eerovaher commented:
> I still think it makes little sense to duplicate `.parse` -- if we go that route, I think we should just give up on the mixin idea, and go back to subclassing instead.

Even in d7f53e74523781e203e8b3ebf568705de1b63e7e all the parsing formatters still implement their own `parse()` methods. My suggestion of keeping the `_do_parse()` method for now would therefore not increase method duplication, but it would keep the diff smaller because we wouldn't have edit `_known_formats()` or `Base.parse()` or the lines that call `_do_parse()`.

eerovaher commented:
> really, `Base` should simply no have a `.parse` method...

`Base.parse()` is a skeleton method that exists to define the public API for parsing unit formatters. In modern Python this could be achieved by defining a `Protocol` and it could be a good idea to introduce protocols for write-only and parsing formatters, but we shouldn't be uppdating the signature of `Base.parse()` until such protocols are defined.

nstarman commented on astropy/units/format/base.py:
```diff
@@ -203,8 +200,135 @@ def _to_string(cls, unit: UnitBase, *, fraction: bool | str) -> str:
+        try:
+            return cls._parser.parse(s, lexer=cls._lexer, debug=debug)
+        except AttributeError:
+            raise NotImplementedError(
+                f"Can not parse with {cls.__name__} format"
+            ) from None
```
```suggestion
            msg = f"Can not parse with {cls.__name__} format"
            raise NotImplementedError(msg) from None
```

mhvk commented:
@nstarman - the goal I am now hoping to go towards is basically become the better version of my option 4 you described in https://github.com/astropy/astropy/pull/17397#pullrequestreview-2438948082, which seems sensible as long as we can call the abstract classes `*Base` for backward compatibility. 

I did not go there directly but went with the mixin mode because @eerovaher suggested we proceed cautiously (likely rightly) and it seemed a good way to move forward, simplifying the structure we have, without precluding the final goal, yet ensuring we do not overly commit to private implementation details. (Also, as is, the mixins are indeed orthogonal.)

Right now, I think we're a bit stuck on a technicality, of whether my solution for `Base.parse()` is good enough for going forward, or whether it is better to keep the `Base` implementation one that just returns `NotImplementedError` and duplicate the actual implementation in all the parsing formats (since they don't inherit from each other any more, they all have to have it separately; it cannot be in the mixin if we want to preserve orthogonality).

Anyway, I'm hoping that you can help decide one way or another. In the meantime, I'm focussing on Quantity 2.0 -- that certainly has been fun, actually writing something from more-or-less scratch! (I think I've got the ufuncs under control now too.)

nstarman commented:
> @nstarman - the goal I am now hoping to go towards is basically become the better version of my option 4 you described in https://github.com/astropy/astropy/pull/17397#pullrequestreview-2438948082, which seems sensible as long as we can call the abstract classes *Base for backward compatibility

👍 

> Right now, I think we're a bit stuck on a technicality, of whether my solution for Base.parse() is good enough for going forward, or whether it is better to keep the Base implementation one that just returns NotImplementedError and duplicate the actual implementation in all the parsing formats

If I'm understanding @eerovaher's point correctly, it's that your current solution for `Base.parse` should be done in a followup PR, for API / refactoring reasons. On the API front I think we could just make this an API-changing PR with a changelog, but for refactoring towards Protocol, etc, I do agree we should probably not change `Base.parse`. So it's probably best to duplicate the implementation.

eerovaher commented:
The question here is not about runtime behavior, it is about specification of public API.

mhvk commented:
OK, I pushed a change that reintroduces `_do_parse()` in the parser mixin, and uses that. I kept the signature identical to what it was, but changed the calls to use `debug=debug` so we can eventually move to a keyword-only version.

eerovaher commented:
> Thanks, both. Interesting to see #17421. Comments on the implementation there, but since it ends up removing `Base.parse()`, it means that if we were to merge that PR before this one, I could just add a `ParsingBase(Base)` that only defines `.parse()` and have the actual implementations inherit from it, without the duplication that I dislike so much.

Are you suggesting that #17421 should be merged first?

mhvk commented:
> Are you suggesting that #17421 should be merged first?

That's no longer necessary - the PRs have become independent as I followed your suggestion of just reintroducing `._do_parse()`.

mhvk commented on astropy/units/format/base.py:
```diff
@@ -208,3 +205,111 @@ def parse(cls, s: str) -> UnitBase:
+    _deprecated_units: ClassVar[frozenset[str]] = frozenset()
+
+    @classmethod
+    def _do_parse(cls, s: str, debug: bool = False) -> UnitBase:
```
Sure, made this keyword only.

mhvk replied:
Actually, I'll go the other way, to avoid doing other methods too.

mhvk commented on astropy/units/format/cds.py:
```diff
@@ -280,7 +280,7 @@ def _format_superscript(cls, number: str) -> str:
     def format_exponential_notation(
         cls, val: UnitScale | np.number, format_spec: str = ".8g"
     ) -> str:
-        return super(Generic, cls).format_exponential_notation(val, format_spec)
+        return super().format_exponential_notation(val, format_spec)
```
Thanks for catching that!

mhvk commented:
> I only have a few small remarks about the final patch, but there should be no trace of `ParsingFormatBase` and `GenericParserBase` in the commit history.

Feel free to squash-merge if you prefer (I think it is actually fine as is, but am fairly sure that editing this to perfection would not be good use of my time (and astropy paying for it) given how unlikely it is we will ever look back at it).

nstarman reviewed (approved):
I'm happy with this! 
Signature cleanup can happen as a followup.

eerovaher commented on astropy/units/format/ogip.py:
```diff
@@ -355,4 +357,10 @@ def to_string(
+
+    @classmethod
+    def format_exponential_notation(
+        cls, val: float | np.number, format_spec: str = "g"
```
I spotted a typing error.
```suggestion
        cls, val: UnitScale | np.number, format_spec: str = "g"
```

eerovaher commented:
> ...given how unlikely it is we will ever look back at it).

Clearly a lot of `astropy` development has taken place under that assumption. Sadly I cannot argue that this assumption is false because it is possible to force it to become true simply by ensuring that the commit history is useless enough.

Squashing the first five commits here wouldn't be too bad, but the commit that updates tests should come first, if it should be kept in this pull request at all. The main reason for updating the tests was to ensure we don't accidentally change an error message. Updating the tests together with code makes that much more difficult to verify (come to think of it, it's also strange to update the tests with this purpose in mind after already having updated the code).

mhvk commented:
OK, corrected the wrong type (must have slipped by in a rebase, thanks for catching that) and squashed the commits that introduced the parser, so that one goes straight to the mixin class. Will enable merge now.
